### PR TITLE
Add RuboCop for `assert_not` over `assert !`

### PR DESCRIPTION
**Pull request by composerinteralia**
We added `assert_not` in f75addd "to replace warty 'assert !foo'".
fa8d35b agrees that it is warty, and so do I. This custom Rubocop rule
turns the wart into a violation.

~~As with my last custom cop, https://github.com/rails/rails/pull/32441,
I want to make sure this looks right on code climate before pushing
another commit to autocorrect everything.~~ seems to have worked as expected

@toshimaru I just noticed
https://github.com/toshimaru/rubocop-rails/pull/26
Needing to copy over all the code for these is definitely not great.
At one point I had thought to put them in a gem.
Is there a better way to add these custom cops, or were you saying we
shouldn't have custom cops at all?

### CODE REVIEW

**Comment by toshimaru** [comment]
@composerinteralia I would say adding those custom cops to RuboCop gem is the better solution. (Not sure if RuboCop team merges it)

Have you tried to add it to RuboCop?

**Comment by composerinteralia** [comment]
I have not tried adding them to RuboCop, but I can try that if it makes sense. Thanks!

File: ci/custom_cops/lib/custom_cops/assert_not.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+  #   assert_not x
+  #
+  class AssertNot < RuboCop::Cop::Cop
+    MSG = "Prefer `assert_not` over `assert !`"
+
+    def_node_matcher :offensive?, "(send nil? :assert (send ... :!))"
+
+    def on_send(node)
+      return unless offensive?(node)
+      add_offense(node, location: :expression, message: MSG)
```
Comment (Edouard-chin) [inline ci/custom_cops/lib/custom_cops/assert_not.rb]: Small nitpick, you don't need the `location: :expression` as well as the message kwarg. https://github.com/bbatsov/rubocop/blob/72bb3922b071712a9262df04e6f5ab5e6f509088/lib/rubocop/cop/internal_affairs/redundant_location_argument.rb#L12-L16
The msg is automatically inferred from the constant

Comment (composerinteralia) [reply ci/custom_cops/lib/custom_cops/assert_not.rb]: Thanks!

**Comment by composerinteralia** [comment]
Oops, my autocorrect method isn't handling parentheses. I will fix that.

File: ci/custom_cops/test/custom_cops/assert_not_test.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+require "support/cop_helper"
+require "./lib/custom_cops/assert_not"
```
Comment (jeremy) [inline ci/custom_cops/test/custom_cops/assert_not_test.rb]: Should we `require_relative …`?

**Review by jeremy — COMMENTED** [review]
Nice! Thanks @composerinteralia.

**Comment by championrajkumar** [comment]
Please unsubscribe me.



On Tue, Apr 17, 2018 at 6:34 PM, Ruby on Rails Bot <notifications@github.com

**Comment by jeremy** [comment]
Hey @championrajkumar, nobody else can unsubscribe you, but you can unsubscribe at the bottom of the email, or click through to the GitHub pull request and hit the Unsubscribe button in the sidebar:
screen shot 2018-04-17 at 23 07 42

File: ci/custom_cops/test/custom_cops/assert_not_test.rb (hunk 2 above)
Comment (composerinteralia) [reply ci/custom_cops/test/custom_cops/assert_not_test.rb]: Done

**Comment by composerinteralia** [comment]
OK, it handles `assert(!x)` with parentheses now as well.

https://github.com/bbatsov/rubocop/pull/5801 would move my previous custom cop into RuboCop. If we end up going in that direction, I can make a PR to RuboCop instead.

**Comment by koic** [comment]
Thank you for mentioning bbatsov/rubocop#5801. So I believe that we can get the criteria to judgment in the near future.

**Pull request merged**